File: Funq/Funq.Tests.Performance/Bench.fs
module Funq.Tests.Performance.Bench
open Funq.Tests.Performance
open System.Diagnostics
open System
open System.Threading
open System.IO
open Funq.Tests
open System.CodeDom.Compiler
open Funq.FSharp.Implementation
open ExtraFunctional
type MutableList<'t> = System.Collections.Generic.List<'t>

let Results = MutableList<MetaContainer>()
let mutable Watch = Stopwatch()
let mutable Timer : Timer = null
let mutable Removes = 5
let mutable Runs = 20
let mutable Timeout = 20000
let mutable RunningThread : Thread = null
let inline (./) num1 num2 = num1 / (float num2)
let mutable OnRun = 
    fun () -> 
        GC.Collect()
        GC.WaitForPendingFinalizers()
    
let invoke (test : unit -> unit)  = 
    let runner() = 
        for i = 0 to Removes do
            test()
        Watch.Reset()
        OnRun()
        Watch.Start()
        for i = 0 to Runs do
            test()
        Watch.Stop()

    let thread = Thread(ThreadStart(runner))
    try
        thread.Start()
        let finished = 
            if Timeout = -1 then
                thread.Join()
                true 
            else 
                thread.Join(Timeout)
        let time = Watch.Elapsed.TotalMilliseconds / (float Runs)
        let time = if finished then Time(time) else TimedOut
        time
    with
        | ex -> 
            Error ex.Message

File: Funq/Funq.Tests.Unit.CSharp/Bench.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Globalization;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Funq.FSharp;

namespace Funq.Tests.Unit.CSharp
{
	public abstract class TimeResult {
		public static implicit operator TimeResult(double d) {
			return new Time(d);
		}

		public static TimeResult TimeOut {
			get {
				return new TimedOut();
			}
		}
	}

	public class Time : TimeResult {
		public double Value;
		public Time(double value) {
			Value = value;
		}

		public override string ToString() {
			return Value.ToString(CultureInfo.InvariantCulture);
		}
	}

	public class TimedOut : TimeResult {
		
	}
	

	internal class Bench {
		public int Drops = 5;
		public int Runs = 10;
		public int MsTimeout = 20000;
		public Stopwatch Watch = new Stopwatch();

		public Action OnRun = () => {
			GC.Collect();
			GC.WaitForPendingFinalizers();
		};

		public TimeResult InvokeTest(Action act) {
			Action runner = () => {
				for (int i = 0; i < Drops; i++) {
					act();
				}
				Watch.Reset();
				OnRun();
				Watch.Start();
				for (int i = 0; i < Runs; i++) {
					act();
				}
				Watch.Stop();
			};

			var thread = new Thread(() => runner());
			thread.Start();
			var succeess = thread.Join(MsTimeout);
			if (succeess) {
				return Watch.Elapsed.TotalMilliseconds/Runs;
			}
			else {
				return new TimedOut();
			}
		}
	}
}
